charms/worker/k8s/src/snap.py

```diff
@@ -269,23 +283,28 @@ def management(charm: K8sCharmProtocol) -> None:
 
     Arguments:
         charm: The charm instance
+
+    Raises:
+        SnapError: when the management issue cannot be resolved
     """
     cache = snap_lib.SnapCache()
     for args in _parse_management_arguments(charm):
-        which = cache[args.name]
+        which: snap_lib.Snap = cache[args.name]
         if block_refresh(which, args, charm.is_upgrade_granted):
             continue
         install_args = args.dict(exclude_none=True)
         if isinstance(args, SnapFileArgument) and which.revision != "x1":
             snap_lib.install_local(**install_args)
-        elif isinstance(args, SnapStoreArgument) and args.revision:
-            if which.revision != args.revision:
-                log.info("Ensuring %s snap revision=%s", args.name, args.revision)
+        elif isinstance(args, SnapStoreArgument):
+            log.info("Ensuring args=%s current=%s", str(args), str(which))
+            new_rev = bool(args.revision) and which.revision != args.revision
+            new_channel = bool(args.channel) and which.channel != args.channel
+            new_state = which.state not in (snap_lib.SnapState.Present, snap_lib.SnapState.Latest)
+            if any([new_rev, new_channel, new_state]):
```

Comment:
I think these changes collapse nicely.  we run `ensure and hold` always regardless of whether we're following a revision or a channel.  Cluster wide upgrades are ill-advised if they are done at the same moment